gdnative-core/src/geom/basis.rs

```diff
@@ -17,4 +20,79 @@ impl Basis {
     pub fn from_sys(c: sys::godot_basis) -> Self {
         unsafe { std::mem::transmute::<sys::godot_basis, Self>(c) }
     }
+
+    pub fn identity() -> Basis {
+        Basis {
+            elements: [
+                Vector3::new(1.0, 0.0, 0.0),
+                Vector3::new(0.0, 1.0, 0.0),
+                Vector3::new(0.0, 0.0, 1.0),
+            ],
+        }
+    }
+
+    pub fn from_diagonal(p_diag: Vector3) -> Basis {
+        Basis {
+            elements: [
+                Vector3::new(p_diag.x, 0.0, 0.0),
+                Vector3::new(0.0, p_diag.y, 0.0),
+                Vector3::new(0.0, 0.0, p_diag.z),
+            ],
+        }
+    }
+
+    // Note: It's ok to use euclid::UnknownUnit as the type for both Src and Dst
+    pub fn from_transform<Src, Dst>(transform: &Transform3D<f32, Src, Dst>) -> Basis {
+        // Note - this encodes just the rotation and scaling
```

Comment:
This information should be in doc comments, because it is significant to users as well as developers:

```suggestion
    /// Creates a `Basis` from the rotation and scaling of the provided transform, in `Dst` space.
    pub fn from_transform<Src, Dst>(transform: &Transform3D<f32, Src, Dst>) -> Basis {
```

- - -

If type inference is a problem, a version without type parameters can be added:

```suggestion
    /// Creates a `Basis` from the rotation and scaling of the provided transform.
    pub fn from_transform(transform: &Transform3D<f32, UnknownUnit, UnknownUnit>) -> Basis {
        Self::from_typed_transform(transform)
    }

    /// Creates a `Basis` from the rotation and scaling of the provided transform, in `Dst` space.
    pub fn from_typed_transform<Src, Dst>(transform: &Transform3D<f32, Src, Dst>) -> Basis {
```

Reply:
If type inference is a problem -- I think that your suggestion will actually prevent users from using typed transforms. They would be forrced only to use untyped ones.

Reply:
> I think that your suggestion will actually prevent users from using typed transforms. They would be forrced only to use untyped ones.

Could you explain why you think that is the case? A `from_typed_transform` method is added in the second version with the same signature as your original method. I can't see how users are prevented from  using typed transforms or "forrced" to only use untyped ones.

I only thought that type inference might have been a problem because you considered "It's ok to use euclid::UnknownUnit as the type for both Src and Dst" a piece of notable information. Normally, that's implied by the type signature: if it's generic and didn't have a trait bound that excluded a type, of course you can use that type with it! Thus, it only makes sense to put emphasis on that information *if* a user will have to turbofish frequently *and* it's not obvious what to put into the angle brackets -- *if* type inference didn't work. If that's not the case, then I'd suggest removing the note.

- - -

```rust
/// Note - a Basis object encodes just the rotation and scaling (not skew)
```

Please replace this with:

```rust
/// Creates a `Basis` that encodes the rotation and scaling of the input transform, in `Dst` space.
```

...since doc-comments should be descriptions of the item. Your version wasn't in the correct tone.